lsp/Main.hs

```diff
@@ -39,29 +42,58 @@ handlers = mconcat
       pure ()
   , requestHandler STextDocumentHover $ \req responder -> do
       let RequestMessage _ _ _ (HoverParams doc pos _workDone) = req
-          Position _l _c' = pos
-          rsp = Hover ms (Just range)
-          ms = HoverContents $ markedUpContent "lsp-demo-simple-server" "Hello world"
-          range = Range pos pos
-      rsp <- lookupToken pos doc
-      -- responder (Right $ Just rsp)
+      let (TextDocumentIdentifier uri) = doc
+      let exceptHover = lookupTokenBare' pos uri
+      runDiagnosticsOnHover uri exceptHover
+      rsp <- runHoverResponse uri exceptHover
       responder (Right rsp)
   , notificationHandler J.STextDocumentDidOpen $ \msg -> do
     let doc  = msg ^. J.params . J.textDocument . J.uri
         fileName =  J.uriToFilePath doc
     liftIO $ debugM "reactor.handle" $ "Processing DidOpenTextDocument for: " ++ show fileName
     for_ fileName $ \fn -> do
-      contents <- liftIO $ readFile fn
+      contents <- liftIO $ readFile fn
       parseAndSendErrors doc $ T.pack contents
   , notificationHandler J.STextDocumentDidSave $ \msg -> do
     let doc  = msg ^. J.params . J.textDocument . J.uri
         fileName =  J.uriToFilePath doc
     liftIO $ debugM "reactor.handle" $ "Processing DidSaveTextDocument for: " ++ show fileName
     for_ fileName $ \fn -> do
-      contents <- liftIO $ readFile fn
+      contents <- liftIO $ readFile fn
       parseAndSendErrors doc $ T.pack contents
   ]
 
+makeDiagErr :: Err -> [Diagnostic]
+makeDiagErr err = [J.Diagnostic
+                  (errorRange err)
+                  (Just J.DsError)  -- severity
+                  Nothing  -- code
+                  (Just "lexer") -- source
+                  (T.pack $ msg err)
+                  Nothing -- tags
+                  (Just (J.List []))
+                ]
+
+publishError :: MonadLsp config m => Uri -> Err -> m ()
+publishError uri err = publishDiagnostics 100 (toNormalizedUri uri) Nothing (partitionBySource $ makeDiagErr err)
+
+clearError :: MonadLsp config m => Uri -> m ()
+clearError uri = publishDiagnostics 100 (toNormalizedUri uri) Nothing (Map.singleton(Just "lexer") (Data.SortedList.toSortedList []))
+
+runDiagnosticsOnHover :: MonadLsp config m => Uri -> ExceptT Err IO a -> m ()
+runDiagnosticsOnHover uri ioHover = do
+  eitherHover <- liftIO $ runExceptT ioHover
+  case eitherHover of
+    Left err -> do
+      publishError uri err
+    Right hover -> do
+      clearError uri
+
+runHoverResponse :: Uri -> ExceptT Err IO Hover -> LspT () IO (Maybe Hover)
+runHoverResponse uri ioHover = do
+  eitherHover <- liftIO $ runExceptT ioHover
```

Comment:
And here it is ran again.

Instead what we should do is to forward the result of it.

This line
```
eitherHover <- liftIO $ runExceptT ioHover
```
should be what's used at this line
```
let exceptHover = lookupTokenBare' pos uri
```
and the two helper functions can take an `Either Err Hover` as their arguments.

(This would also reduce code duplication)

Reply:
Agree, I was thinking about something like this but couldn't figure out how to do it.

Reply:
An alternative would be to merge the two helpers into a single function:

```
runAndSendDiagnostics :: MonadLsp config m => Uri -> ExceptT Err IO a -> m (Maybe a)
runAndSendDiagnostics uri ioHover = do
  eitherHover <- liftIO $ runExceptT ioHover
  case eitherHover of
    Left err -> do
      publishError uri err
      pure Nothing
    Right hover -> do
      clearError uri
      pure $ Just hover
```
Now this function is responsible for handling the errors, which is indicated by changing the `Either Err` to `Maybe` in the type.

Or doing both changes:
```
sendDiagnosticsOnLeft :: MonadLsp config m => Uri -> Either Err -> m (Maybe a)
sendDiagnosticsOnLeft uri eitherHover = do
  case eitherHover of
    Left err -> do
      publishError uri err
      pure Nothing
    Right hover -> do
      clearError uri
      pure $ Just hover
```
or more compactly:
```
sendDiagnosticsOnLeft :: MonadLsp config m => Uri -> Either Err -> m (Maybe a)
sendDiagnosticsOnLeft uri = either ((Nothing <$) . publishError uri) (fmap Just . clearError)
```

But maybe that just makes it more complicated?

Reply:
Actually yes.
I recalled now why I split them - I wanted to get rid of case but couldn't figure out how.

Reply:
My last function doesn't actually work. I should have tested it before writing it.